### PR TITLE
7 feature 7 getgamestate use case class

### PR DESCRIPTION
**Pull request by SMateen21**
Created a GameState class which observes a given Game. It contains information needed for the User to see in the UI such as the current Player and their Cards, the last Card played, the rest of the Players in the Game and their number of Cards, and whether or not the Game has a winner. These attributes are updated whenever Game notifies it's observers, this logic is to be implemented in the interactor. This class comes with full documentation and a test class. 
One thing that may be added here in the future is if the player can pick up a Card, however this logic I believe should be delegated to the interactor.

### CODE REVIEW

File: src/test/java/use_cases/GameStateTest.java
Comment (nyan-ko) [inline src/test/java/use_cases/GameStateTest.java:1]: add access modifiers for the test methods

File: src/main/java/use_cases/GameState.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+     * current Card, their Cards, the other players and the number of Cards they have,
+     * as well as if the Game has a winner.
+     * @param game The Game containing the information to update this GameState.
+     */
+    public GameState(ObservableGame game) {
+
+        this.currentPlayer = game.getCurrentTurn();
+        this.currentCard = game.getCurrentCard();
+        for (Player player: game.getPlayers()) {
+            if(player != currentPlayer) {
```
Comment (nyan-ko) [inline src/main/java/use_cases/GameState.java:33]: what is the reasoning behind keeping the current player out of the playersAndCards hashmap? even if it is supposed to store opponents, i believe it's easier to store every player and exclude the current player

File: src/main/java/use_cases/GameState.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        this.currentPlayer = game.getCurrentTurn();
+        this.currentCard = game.getCurrentCard();
+        this.hasWinner = game.hasWinner();
+        for(Player p: game.getPlayers()) {
+            if(p.equals(currentPlayer)) {
+                this.playersAndCards.remove(p);
+            }
+            else {
+                this.playersAndCards.remove(p);
+                this.playersAndCards.put(p, p.getNumCards());
```
Comment (meiiph) [inline src/main/java/use_cases/GameState.java:57]: This for loop may be redundant, you could consider rewriting it.

File: src/test/java/use_cases/GameStateTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    private Player p1;
+    private Player p2;
+    private ArrayList<Player> players;
+    private GameState gameState;
+
+    /**
+     * Initialize all objects needed to test GameState.
+     */
+    @BeforeEach
+    void setUp(){
```
Comment (meiiph) [inline src/test/java/use_cases/GameStateTest.java:28]: Access modifier?

File: src/test/java/use_cases/GameStateTest.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        game = new Game(deck, players);
+        gameState = new GameState(game);
+
+    }
+
+    /**
+     * Set all objects created in setUp() to null to free memory.
+     */
+    @AfterEach
+    void tearDown() {
```
Comment (meiiph) [inline src/test/java/use_cases/GameStateTest.java:52]: Access modifier?

File: src/test/java/use_cases/GameStateTest.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        game = null;
+        players = null;
+    }
+
+    /**
+     * Test whether GameState is properly updated when updateGameObserver is called.
+     * This test method will also test the getter methods in GameState as well.
+     */
+    @Test
+    void testUpdateGameObserver() {
```
Comment (meiiph) [inline src/test/java/use_cases/GameStateTest.java:65]: Access modifier?

File: src/main/java/use_cases/GameState.java (hunk 2 above)
Comment (nyan-ko) [inline src/main/java/use_cases/GameState.java:57]: i believe the HashMap.replace method would work better here

**Review by meiiph — CHANGES_REQUESTED** [review]
Looks good, just waiting for the changes I added in the comments on the code. :)

File: src/main/java/use_cases/GameState.java (hunk 1 above)
Comment (SMateen21) [reply src/main/java/use_cases/GameState.java:33]: The logic of excluding would have to be delegated to lower level modules, I thought it best to contain most of the logic in the use case layer.

File: src/main/java/use_cases/GameState.java (hunk 2 above)
Comment (SMateen21) [reply src/main/java/use_cases/GameState.java]: updated the code so that this is no longer required.

**Review by meiiph — APPROVED** [review]
👍

**Review by nyan-ko — APPROVED** [review]
lgtm 👍

**Review by lilyxmeng — APPROVED** [review]
Looks good

**Pull request merged**